cli/command/stack/kubernetes/watcher.go

```diff
@@ -1,116 +1,253 @@
+				status.podsPending++
+			case apiv1.PodRunning:
+				status.podsRunning++
+			case apiv1.PodSucceeded:
+				status.podsSucceeded++
+			case apiv1.PodFailed:
+				status.podsFailed++
+			case apiv1.PodUnknown:
+				status.podsUnknown++
+			}
+			if podutils.IsPodReady(pod) {
+				status.podsReady++
 			}
 		}
-
-		if !ready {
-			return false
-		}
-
-		serviceName := pod.Labels[labels.ForServiceName]
-		serviceUp[serviceName] = true
 	}
+	oldStatus := pw.services[serviceName]
+	if oldStatus != status {
+		pw.statusUpdates <- status
+	}
+	pw.services[serviceName] = status
+}
 
-	for _, serviceName := range serviceNames {
-		if !serviceUp[serviceName] {
+func (pw *podWatcher) allReady() bool {
+	for _, status := range pw.services {
+		if status.podsReady == 0 {
 			return false
 		}
 	}
-
 	return true
 }
 
-func timeTimes(n int32) string {
-	if n == 1 {
-		return "time"
+func (pw *podWatcher) OnAdd(obj interface{}) {
+	pw.handlePod(obj)
+}
+
+func (pw *podWatcher) OnUpdate(oldObj, newObj interface{}) {
+	pw.handlePod(newObj)
+}
+
+func (pw *podWatcher) OnDelete(obj interface{}) {
+	pw.handlePod(obj)
+}
+
+func (w *deployWatcher) waitForPods(ctx context.Context, stackName string, serviceNames []string, e chan error, statusUpdates chan serviceStatus) {
+	informer := newPodInformer(w.pods, stackName, cache.Indexers{
+		"byservice": func(obj interface{}) ([]string, error) {
+			pod, ok := obj.(*apiv1.Pod)
+			if !ok {
+				return nil, errors.New("pod has an unexpected type")
+			}
+			return []string{pod.Labels[labels.ForServiceName]}, nil
+		}})
+	services := map[string]serviceStatus{}
+	for _, name := range serviceNames {
+		services[name] = serviceStatus{name: name}
+	}
+	pw := &podWatcher{
+		stackName:     stackName,
+		services:      services,
+		resultChan:    e,
+		starts:        map[string]int32{},
+		indexer:       informer.GetIndexer(),
+		statusUpdates: statusUpdates,
 	}
+	informer.AddEventHandler(pw)
+	informer.Run(ctx.Done())
+}
+
+func newPodInformer(podsClient podListWatch, stackName string, indexers cache.Indexers) cache.SharedIndexInformer {
+	return cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.LabelSelector = labels.SelectorForStack(stackName)
+				options.IncludeUninitialized = true
+				return podsClient.List(options)
+			},
+
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = labels.SelectorForStack(stackName)
+				options.IncludeUninitialized = true
+				return podsClient.Watch(options)
+			},
+		},
+		&apiv1.Pod{},
+		time.Second*5,
+		indexers,
+	)
+}
 
-	return "times"
+func newStackInformer(stacksClient stackListWatch, stackName string) cache.SharedInformer {
+	return cache.NewSharedInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.LabelSelector = "com.docker.stack.namespace=" + stackName
```

Comment:
Looks like we have defined the name for this label as a constant in two locations;

https://github.com/docker/cli/blob/236a84759af68d2f2a7d216f0708152a896229ac/cli/compose/convert/compose.go#L13-L16

and

https://github.com/docker/cli/blob/7ad30360c823c4f7748e1ac7ef98ec21414cb691/kubernetes/labels/labels.go#L8-L15

Can you use a constant here? (Also, we should discuss unifying those constants, but not for this PR)